https://towardsdatascience.com/step-by-step-exploratory-data-analysis-on-stroke-dataset-840aefea8739

In [2]:
import pandas as pd
import matplotlib.pyplot as plot
import seaborn as sns
import numpy as np

In [3]:
df = pd.read_csv("data/stroke.csv", encoding="utf-8")
df2 = pd.read_csv('data/train_strokes.csv', encoding='utf-8')
df.head()

id  gender   age  hypertension  heart_disease ever_married  \
0   9046    Male  67.0             0              1          Yes   
1  51676  Female  61.0             0              0          Yes   
2  31112    Male  80.0             0              1          Yes   
3  60182  Female  49.0             0              0          Yes   
4   1665  Female  79.0             1              0          Yes   

       work_type Residence_type  avg_glucose_level   bmi   smoking_status  \
0        Private          Urban             228.69  36.6  formerly smoked   
1  Self-employed          Rural             202.21   NaN     never smoked   
2        Private          Rural             105.92  32.5     never smoked   
3        Private          Urban             171.23  34.4           smokes   
4  Self-employed          Rural             174.12  24.0     never smoked   

   stroke  
0       1  
1       1  
2       1  
3       1  
4       1

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5110 entries, 0 to 5109
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 5110 non-null   int64  
 1   gender             5110 non-null   object 
 2   age                5110 non-null   float64
 3   hypertension       5110 non-null   int64  
 4   heart_disease      5110 non-null   int64  
 5   ever_married       5110 non-null   object 
 6   work_type          5110 non-null   object 
 7   Residence_type     5110 non-null   object 
 8   avg_glucose_level  5110 non-null   float64
 9   bmi                4909 non-null   float64
 10  smoking_status     5110 non-null   object 
 11  stroke             5110 non-null   int64  
dtypes: float64(3), int64(4), object(5)
memory usage: 479.2+ KB


In [5]:
df['stroke'].value_counts()

0    4861
1     249
Name: stroke, dtype: int64

Bộ dữ liệu này ta thấy rõ sự mất cân bằng. Chỉ có 249 bệnh nhân đã từng trải qua đột quỵ trong khi 4861 bệnh nhân không bị.

# Data preprocessing

## 1. ID attribute
Dùng định danh các bệnh nhân, không có nhiều ý nghĩa nên ta xóa cả cột này đi.

In [6]:
df.drop(columns=['id'], inplace=True)

## 2. BMI attribute
Có 40 bệnh nhân không có chỉ số BMI nhưng đã từng đột quỵ (so với 249 bệnh nhân ở trên, tỉ lệ khá lớn) nên với các giá trị NaN này ta sẽ cho bằng chỉ số trung bình BMI

In [7]:
df[df['bmi'].isna() & df['stroke'] == 1]

gender    age  hypertension  heart_disease ever_married      work_type  \
1    Female  61.00             0              0          Yes  Self-employed   
8    Female  59.00             0              0          Yes        Private   
13     Male  78.00             0              1          Yes        Private   
19     Male  57.00             0              1           No       Govt_job   
27     Male  58.00             0              0          Yes        Private   
29     Male  59.00             0              0          Yes        Private   
43   Female  63.00             0              0          Yes        Private   
46   Female  75.00             0              1           No  Self-employed   
50   Female  76.00             0              0           No        Private   
51     Male  78.00             1              0          Yes        Private   
54   Female  63.00             0              0          Yes       Govt_job   
57     Male  78.00             0              0          Yes        Private   
64     Male  75.00             0              0          Yes        Private   
70   Female  76.00             0              0          Yes       Govt_job   
78   Female  51.00             0              0          Yes        Private   
81   Female  66.00             0              0          Yes  Self-employed   
84     Male  58.00             0              0          Yes        Private   
105    Male  58.00             0              0          Yes        Private   
112  Female  76.00             0              0          Yes  Self-employed   
124  Female  72.00             0              0          Yes        Private   
126    Male  78.00             1              0          Yes  Self-employed   
129  Female  75.00             0              0          Yes       Govt_job   
133  Female  38.00             0              0          Yes        Private   
146    Male  65.00             0              0          Yes  Self-employed   
150  Female  79.00             0              0          Yes        Private   
160  Female  76.00             0              0          Yes        Private   
161    Male  71.00             0              1          Yes        Private   
162  Female   1.32             0              0           No       children   
167    Male  79.00             1              0          Yes        Private   
170    Male  64.00             0              0          Yes  Self-employed   
171  Female  79.00             1              1           No  Self-employed   
174  Female  78.00             0              0          Yes  Self-employed   
178  Female  80.00             0              0          Yes       Govt_job   
183  Female  77.00             0              0           No        Private   
189    Male  61.00             0              1          Yes        Private   
198    Male  79.00             0              0          Yes        Private   
200    Male  74.00             0              0          Yes        Private   
218  Female  76.00             1              1          Yes  Self-employed   
227    Male  74.00             0              0          Yes  Self-employed   
247    Male  71.00             1              0          Yes  Self-employed   

    Residence_type  avg_glucose_level  bmi   smoking_status  stroke  
1            Rural             202.21  NaN     never smoked       1  
8            Rural              76.15  NaN          Unknown       1  
13           Urban             219.84  NaN          Unknown       1  
19           Urban             217.08  NaN          Unknown       1  
27           Rural             189.84  NaN          Unknown       1  
29           Rural             211.78  NaN  formerly smoked       1  
43           Urban              90.90  NaN  formerly smoked       1  
46           Urban             109.78  NaN          Unknown       1  
50           Urban              89.96  NaN          Unknown       1  
51           Urban              75.32  NaN  formerly smoked      

In [16]:
df['bmi'].fillna(np.round(df['bmi'].mean(), 1), inplace=True)

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5110 entries, 0 to 5109
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   gender             5110 non-null   object 
 1   age                5110 non-null   float64
 2   hypertension       5110 non-null   int64  
 3   heart_disease      5110 non-null   int64  
 4   ever_married       5110 non-null   object 
 5   work_type          5110 non-null   object 
 6   Residence_type     5110 non-null   object 
 7   avg_glucose_level  5110 non-null   float64
 8   bmi                5110 non-null   float64
 9   smoking_status     5110 non-null   object 
 10  stroke             5110 non-null   int64  
dtypes: float64(3), int64(3), object(5)
memory usage: 439.3+ KB


## 3. Smoking Status attribute <br/>
In addition, 13,292 records or about 30.6% of the dataset had missing values in smoking status feature (from Fig. 1). It was a huge proportion of the dataset. As such, a new category named “not known” was created to account for all these records, rather than dropping them altogether.

In [12]:
print(df['smoking_status'].value_counts())

never smoked       1892
Unknown            1544
formerly smoked     885
smokes              789
Name: smoking_status, dtype: int64


## 4. Gender attribute <br/>

In [19]:
print(df['gender'].value_counts())

Female    2994
Male      2115
Other        1
Name: gender, dtype: int64


In [20]:
df = df[df['gender'] != 'Other']

## 5. Normalize numerical attributes
In this dataset, there are 3 numerical attributes, i.e. age, average glucose level and bmi. Let’s normalize them to ensure that they have equal weightage when building a classifier

In [21]:
# Create a new column for normalized age
df['age_norm']=(df['age']-df['age'].min())/(df['age'].max()-df['age'].min())

# Create a new column for normalized avg glucose level
df['avg_glucose_level_norm']=(df['avg_glucose_level']-df['avg_glucose_level'].min())/(df['avg_glucose_level'].max()-df['avg_glucose_level'].min())

# Create a new column for normalized bmi
df['bmi_norm']=(df['bmi']-df['bmi'].min())/(df['bmi'].max()-df['bmi'].min())

<ipython-input-21-dadf6a7a71b7>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['age_norm']=(df['age']-df['age'].min())/(df['age'].max()-df['age'].min())
<ipython-input-21-dadf6a7a71b7>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['avg_glucose_level_norm']=(df['avg_glucose_level']-df['avg_glucose_level'].min())/(df['avg_glucose_level'].max()-df['avg_glucose_level'].min())
<ipython-input-21-dadf6a7a71b7>:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5109 entries, 0 to 5109
Data columns (total 14 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   gender                  5109 non-null   object 
 1   age                     5109 non-null   float64
 2   hypertension            5109 non-null   int64  
 3   heart_disease           5109 non-null   int64  
 4   ever_married            5109 non-null   object 
 5   work_type               5109 non-null   object 
 6   Residence_type          5109 non-null   object 
 7   avg_glucose_level       5109 non-null   float64
 8   bmi                     5109 non-null   float64
 9   smoking_status          5109 non-null   object 
 10  stroke                  5109 non-null   int64  
 11  age_norm                5109 non-null   float64
 12  avg_glucose_level_norm  5109 non-null   float64
 13  bmi_norm                5109 non-null   float64
dtypes: float64(6), int64(3), object(5)
memor